README.md
## **What this project does**
The climatelocator API returns climate data from different locations and occured disasters related to climate change in that location.

<br/> 


### A location contains the following information:
 * The avarage monthly temperatures in degrees
 * The name of the location
 * The geohash for that location
 

### A disaster contains the following information: the amount of deaths, missing
 * Type of the disaster (hurrican, drought, flood, wildfire, heat wave)
 * The name of the disaster
 * The amount of people injured
 * Missing
 * Injuries
 * The amount of damage in dollars

<br/>

**See getting started for possible endpoints**
#
## **Why is it useful**
This API can be used to visualize weather and changes in weather, as well as the climate disasters that occured at these locations.

By using this data, there is the possibility to create datavisualisations or simply display any weather information on a website. Furthermore, the data can be used for installations and real-time applications. 
#
## **Getting started**

1. Clone the repo
```
git clone https://github.com/eliseboes/climatelocator.git
```
2. Run docker
```
docker-compose up
```
**NOTE:** Make sure that Docker is running on your computer.
#
## **Where can I find help**
Please post your questions and any others in the issue section on Github, with the correct tags.

>bug - Something isn't working

>documentation  Improvements or additions to documentation

>duplicate - This issue or pull request already exists

>enhancement - New feature or request

>good first issue - Good for newcomers

>help wanted - Extra attention is needed

>invalid - This doesn't seem right

>question - Further information is requested

For any questions not related to the project, please contact me at elise.boes@student.ehb.be
#

## **Project status**
In development.

#

## **Git flow**
Git Flow was used in this project, due to the possible future complexity of the API. Information provided in this project is limited, however, can be expanded by providing more tables of data related to climate change.

When extending the API, the project could take up a large scale. Therefore, it is required to correctly Fork the project on another branch. 
**See: 'How to use'**
#
## **Authors**
This project was created by Elise Boes, student Multimedia and Creative Technologies.

CHANGELOG.md
# V0.1.0

## Core
```
[POST] /addlocation endpoint
[POST] /updatelocation endpoint
[DELETE] /removelocation endpoint
[GET] /getlocation:uuid endpoint

[POST] /adddisaster endpoint
[POST] /updatedisaster endpoint
[DELETE] /removedisaster:uuid endpoint
[GET] /disasterbytype:type endpoint
[GET] /getalldisasters endpoint
```

api/src/__tests__/helpers.test.js
const Helpers = require('../utils/helpers.js');

describe('check', () => {
  test('test', () =>{ expect('test').toBe('test')});

  test('if monthly average is an object', () => {
    expect(Helpers.checkYearlyAverages({Jan: 5.1, Feb: 5.8, Mar: 8.6, Apr: 10, May: 11, Jun: 12, Jul: 13, Aug: 14, Sep: 15, Oct: 12, Nov: 11, Dec: 7})).toStrictEqual({Jan: 5.1, Feb: 5.8, Mar: 8.6, Apr: 10, May: 11, Jun: 12, Jul: 13, Aug: 14, Sep: 15, Oct: 12, Nov: 11, Dec: 7});
  });
  test('if monthly average is not an object', () => {
    expect(Helpers.checkYearlyAverages([5.1, 5.8, 8.6, 10, 11, 12, 13, 14, 15, 12, 11, 7])).toBeFalsy();
  });

  test('if geohash has correct length', () => {
    expect(Helpers.checkGeohashLength('gkp')).toStrictEqual('gkp');
  });
  test('if geohash does not have correct length', () => {
    expect(Helpers.checkGeohashLength('gkpgkpebp6mb4')).toBeFalsy();
  });

  test('if geohash contains only letters and/or digits', () => {
    expect(Helpers.checkGeohashFormat('gkp15')).toStrictEqual('gkp15');
  });
  test('if geohash does not contain only letters and/or digits', () => {
    expect(Helpers.checkGeohashFormat('gkpgkpe_%4')).toBeFalsy();
  });

});

api/src/__tests__/integration.test.js
const supertest = require('supertest');
const Helpers = require('../utils/helpers.js');
const app = require('./../server.js');

const request = supertest(app);

// describe('POST /addlocation endpoint', () => {
//     test('check if /addlocation responds to 200 and inserts a location in db', async (done) => {
//         try {
//             await request.post('/addlocation')
//                 .expect(200)
//                 .then((res) => {
//                     done()
//                 });
//         } catch (e) {
//             if (e) console.log(e);
//         }
//     });
// });

api/src/__tests__/integration/end-to-end.test.js
const supertest = require('supertest');
const app = require('../../server.js');
const request = supertest(app);
const Helpers = require('../../utils/helpers.js');

describe('DB connection test', () => {
            test('full connection test', async () => {
                    try {
                        let uuid = await request.post('/locations')
                            .send({
                                    uuid: Helpers.generateUUID(),
                                    name: 'Jamaica',
                                    geohash: 's7jjkf4x5tst',
                                    yearly_averages_low: {
                                        Jan: 2.0,
                                        Feb: 2.0,
                                        Mar: 5.0,
                                        Apr: 10.0,
                                        May: 14.0,
                                        Jun: 18.0,
                                        Jul: 21.8,
                                        Aug: 23.0,
                                        Sep: 20.0,
                                        Oct: 15.0,
                                        Nov: 9.0,
                                        Dec: 4.0
                                    },
                                    yearly_averages_high: {
                                        Jan: 8.0,
                                        Feb: 9.0,
                                        Mar: 12.0,
                                        Apr: 17.0,
                                        May: 21.0,
                                        Jun: 25.5,
                                        Jul: 28.0,
                                        Aug: 29.0,
                                        Sep: 26.0,
                                        Oct: 20.0,
                                        Nov: 15.0,
                                        Dec: 11.0
                                    }
                                })
                                .expect(201)
                                .then((response) => {
                                    const parsed = JSON.parse(response.text)
                                    return parsed.res[0].uuid
                                }).catch((error) => {
                                    console.log(error)
                                });
                                const getRequest = await request.get(`/locations/${uuid}`).expect(200)
                                const requestParse = JSON.parse(getRequest.text)
                                expect(requestParse.res).toHaveLength(1)
                            }
                        catch (error) {
                            throw error
                        }
                    })
            })

api/src/utils/helpers.js

const {v1: uuidv1 } = require('uuid');

const Helpers = {
  generateUUID: () => {
     const uuid = uuidv1();  
     return uuid;
  },
  /** * Checks type op yearly averages param *
   * @param months - object 'months' where type of has to be checken *
   * @returns months if correct, false if not correct */

  checkYearlyAverages: (months) =>{
    if(typeof months === 'object' && months !== null && !Array.isArray(months)){
      return months
    }else{
      return false
    }
  },
  
  /** * Checks length of geohash *
   * @param geohash - geohash of which the length has to be checked *
   * @returns geohash if correct, false if not correct */

  checkGeohashLength: (geohash) =>{
    if(geohash.length <= 12){
      return geohash
    }else{
      return false
    }
  },

   /** * Checks  if geohash only contains digits and numbers *
   * @param geohash - geohash of which the format has to be checked *
   * @returns geohash if correct, false if not correct */
  
  checkGeohashFormat: (geohash) =>{
    if(geohash.match("^[A-​Za-z0-9]+$") !== null){
      return geohash
    }else{
      return false
    }
  }

}

module.exports = Helpers

api/src/utils/DatabaseHelper.js
// const initialiseTables = async() => {
//   const pg = require('knex')({
//     client: 'pg',
//     version: '9.6',
//     searchPath: ['knex', 'public'],
//     connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/climatelocator'
//   });

//   await pg.schema.hasTable('year').then(async (exists) => {
//     if (!exists) {
//       await pg.schema
//         .createTable('year', (table) => {
//           table.increments();
//           table.uuid('uuid');
//           table.string('most_affected_country');
//           table.string('hottest_temperature');
//           table.string('num_disasters');
//           table.timestamps(true, true);
//         })
//         .then(async () => {
//           console.log('created table year');
//         });

//     }
//   });
//   await pg.schema.hasTable('locations').then(async (exists) => {
//     if (!exists) {
//       await pg.schema
//         .createTable('locations', (table) => {
//           table.increments();
//           table.uuid('uuid');
//           table.string('name');
//           table.string('geohash');
//           table.string('yearly_averages');
//           table.string('year');
//           table.string('year_id');
//           table.timestamps(true, true);
//         })
//         .then(async () => {
//           console.log('created table locations');
//         });
//     }
//   });
// }

// module.exports = initialiseTables;

api/src/server.js
const express = require('express')
const bodyParser = require('body-parser');
const http = require('http');
const Helpers = require('./utils/helpers.js');
const {
  generateUUID
} = require('./utils/helpers.js');
const {
  doesNotMatch
} = require('assert');
const dbHelper = require('./utils/DatabaseHelper.js');

const port = 3000

const app = express();
http.Server(app);


app.use(bodyParser.json());
app.use(
  bodyParser.urlencoded({
    // to support URL-encoded bodies
    extended: true
  })
);

/*--------------------- INFO ----------------------*/

app.get('/', async (req, res) => {
  res.send(
    '<h2>Endpoints</h2>'
  );
});

/*---------------- ADD LOCATION ------------------*/

const pg = require('knex')({
  client: 'pg',
  version: '9.6',
  searchPath: ['knex', 'public'],
  connection: process.env.PG_CONNECTION_STRING ? process.env.PG_CONNECTION_STRING : 'postgres://example:example@localhost:5432/climatelocator'
});

async function initialiseTables() {
  await pg.schema.hasTable('disasters').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('disasters', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('name');
          table.string('type');
          table.string('fatalities');
          table.string('injuries');
          table.string('missing');
          table.string('damage');
          table.timestamps(true, true);
        })
        .then(async () => {
          let disasters = [{
              uuid: Helpers.generateUUID(),
              name: 'Hurricane Eta',
              type: 'hurricane',
              fatalities: 211,
              missing: 120,
              damage: 7900000000
            },
            {
              uuid: Helpers.generateUUID(),
              name: '2020 East Africa Floods',
              type: 'flood',
              fatalities: 453,
              missing: 8
            }, {
              uuid: Helpers.generateUUID(),
              name: '2019 European Heat Wave',
              type: 'heat wave',
              fatalities: 869
            },
            {
              uuid: Helpers.generateUUID(),
              name: 'Typhoon Hagibis',
              type: 'hurricane',
              fatalities: 98,
              missing: 7,
              damage: 15000000000
            }
          ]
          console.log('created table disasters');
          for (let i = 0; i < disasters.length; i++) {
            await pg.table('disasters').insert(disasters[i]);
          }
        });
    }
  });
  await pg.schema.hasTable('locations').then(async (exists) => {
    if (!exists) {
      await pg.schema
        .createTable('locations', (table) => {
          table.increments();
          table.uuid('uuid');
          table.string('name');
          table.string('geohash');
          table.string('yearly_averages_low');
          table.string('yearly_averages_high');
          table.string('disaster_id');
          table.timestamps(true, true);
        })
        .then(async () => {
          console.log('created table locations');
          let locations = [{
              uuid: Helpers.generateUUID(),
              name: 'Kenya',
              geohash: 'sb4cn8hn24buk',
              yearly_averages_low: {
                Jan: 2.0,
                Feb: 2.0,
                Mar: 5.0,
                Apr: 10.0,
                May: 14.0,
                Jun: 18.0,
                Jul: 21.8,
                Aug: 23.0,
                Sep: 20.0,
                Oct: 15.0,
                Nov: 9.0,
                Dec: 4.0
              },
              yearly_averages_high: {
                Jan: 8.0,
                Feb: 9.0,
                Mar: 12.0,
                Apr: 17.0,
                May: 21.0,
                Jun: 25.5,
                Jul: 28.0,
                Aug: 29.0,
                Sep: 26.0,
                Oct: 20.0,
                Nov: 15.0,
                Dec: 11.0
              },
              disaster_id: 'a34a0ac0-4dda-11eb-b21e-6504195ef07a'
            },
            {
              uuid: Helpers.generateUUID(),
              name: 'Paris',
              geohash: 'u09tvmqrep',
              yearly_averages_low: {
                Jan: 2.0,
                Feb: 2.0,
                Mar: 5.0,
                Apr: 10.0,
                May: 14.0,
                Jun: 18.0,
                Jul: 21.8,
                Aug: 23.0,
                Sep: 20.0,
                Oct: 15.0,
                Nov: 9.0,
                Dec: 4.0
              },
              yearly_averages_high: {
                Jan: 8.0,
                Feb: 9.0,
                Mar: 12.0,
                Apr: 17.0,
                May: 21.0,
                Jun: 25.5,
                Jul: 28.0,
                Aug: 29.0,
                Sep: 26.0,
                Oct: 20.0,
                Nov: 15.0,
                Dec: 11.0
              },
              disaster_id: 'a34a0ac1-4dda-11eb-b21e-6504195ef07a'
            }, {
              uuid: Helpers.generateUUID(),
              name: 'Tokyo',
              geohash: 'xn76cydhz',
              yearly_averages_low: {
                Jan: 2.0,
                Feb: 2.0,
                Mar: 5.0,
                Apr: 10.0,
                May: 14.0,
                Jun: 18.0,
                Jul: 21.8,
                Aug: 23.0,
                Sep: 20.0,
                Oct: 15.0,
                Nov: 9.0,
                Dec: 4.0
              },
              yearly_averages_high: {
                Jan: 8.0,
                Feb: 9.0,
                Mar: 12.0,
                Apr: 17.0,
                May: 21.0,
                Jun: 25.5,
                Jul: 28.0,
                Aug: 29.0,
                Sep: 26.0,
                Oct: 20.0,
                Nov: 15.0,
                Dec: 11.0
              },
              disaster_id: 'a34a0ac1-4dda-11eb-b21e-6504195ef07a'
            }
          ]
          for (let i = 0; i < locations.length; i++) {
            await pg.table('locations').insert(locations[i]);
          }
        });
    }
  });
}
initialiseTables();

/**  add location
 * @params req.body 
 * @returns status 201 and inserted location when OK, status 404 when not OK
 */
app.post('/locations', async (req, res) => {
  const data = req.body;
  if (Helpers.checkGeohashFormat(data.geohash) == data.geohash && Helpers.checkGeohashLength(data.geohash) == data.geohash) {
    const result = pg('locations')
      .insert(data)
      .returning('*')
      .then(function (result) {
        res.status(201)
        res.json({
            res: result
          })
          .send();
      }).catch((e) => {
        console.log(e);
        res.status(404).send();
      });
  }
});

/**  delete location
 * @params uuid  
 * @returns status 200 when OK, status 404 when not OK
 */
app.delete('/locations/:uuid', async (req, res) => {
  const uuid = req.params.uuid
  pg('locations')
    .del()
    .where({
      uuid: uuid
    })
    .then(function (result) {
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  update location
 * @params req.body.uuid
 * @returns status 200 and updated location when OK, status 404 when not OK
 */
app.put('/locations', async (req, res) => {
  const uuid = req.body.uuid;
  pg('locations')
    .where({
      uuid: uuid
    })
    .update({
      yearly_averages_high: {
        Jan: 8.0,
        Feb: 10.0,
        Mar: 13.0,
        Apr: 17.0,
        May: 21.0,
        Jun: 25.5,
        Jul: 28.0,
        Aug: 29.0,
        Sep: 26.0,
        Oct: 20.0,
        Nov: 15.0,
        Dec: 11.0
      }
    })
    .then(function (result) {
      res.status(200)
      res.json({
          res: result
        })
        .send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  get location by uuid
 * @params uuid 
 * @returns status 200 and location when OK, status 404 when not OK
 */
app.get('/locations/:uuid', async (req, res) => {
  pg('locations')
    .where({
      uuid: req.params.uuid
    })
    .then(result => {
      res.json({
        res: result
      })
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  get disaster by type
 * @params type
 * @returns status 200 and disasters of selected type when OK, status 404 when not OK
 */
app.get('/disasters/:type', async (req, res) => {
  pg('disasters')
    .where({
      type: req.params.type
    })
    .then(result => {
      res.json({
        res: result
      })
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  get all disasters
 * @params 
 * @returns status 200 and all disasters when OK, status 404 when not OK
 */
app.get('/alldisasters', async (req, res) => {
  pg.select('*')
    .from('disasters')
    .then(result => {
      res.json({
        res: result
      })
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  update disaster by uuid
 * @params uuid  
 * @returns status 200 and updated disaster when OK, status 404 when not OK
 */
app.put('/disasters/:uuid', async (req, res) => {
  const uuid = req.params.uuid;
  pg('disasters')
    .where({
      uuid: uuid
    })
    .update({
      type: 'wildfire'
    })
    .then(function (result) {
      res.json({
        res: result
      })
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/**  delete disaster by uuid
 * @params uuid  
 * @returns status 200 when OK, status 404 when not OK
 */
app.delete('/disasters/:uuid', async (req, res) => {
  const uuid = req.params.uuid;
  pg('disasters')
    .where({
      uuid: uuid
    })
    .del()
    .then(function (result) {
      res.status(200).send();
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});

/** add disaster
 * @params uuid  
 * @returns status 200 and inserted disaster when OK, status 404 when not OK
 */
app.post('/disasters', async (req, res) => {
  const data = req.body;
  pg('disasters')
    .insert(data)
    .returning('*')
    .then(function (result) {
      res.json({
        res: result
      })
      res.status(201).send();
      app.get()
    }).catch((e) => {
      console.log(e);
      res.status(404).send();
    });
});


module.exports = app;
